### DISCUSSION TITLE
How does "like" in a data provider work

### DISCUSSION THREAD

**Discussion by PGLange** [Budibase/budibase#1874] [Help] [answered]
Hi, 
I was using the Feature like to figure out if an email adress is in a field that's containg two email adresses. Initially it worked - but it doesn't now.
![like1](https://user-images.githubusercontent.com/86222990/123818896-a6b0e800-d8f9-11eb-8707-6a83dca9fc15.PNG)
![like2](https://user-images.githubusercontent.com/86222990/123818900-a6b0e800-d8f9-11eb-8554-e186adfd12dd.PNG)

The like is used in a data provider. 

I don't understand how your "like" works

Please help me, 'cause this breaks my app

**Comment by mike12345567** [comment, marked as answer]
Hi @PGLange - we have actually discussed removing the `like` operator as it is quite confusing, it specifically is a fuzzy search, so for example if you had an entry in your database of "Banana" then the like search would match with mis-spellings, like "Fanana". This isn't particularly useful as "Bana" will not match with like as it doesn't perform any sort of starts with or contains search.

The best search for any sort of string lookup is "starts with", we realise there will be some cases where this isn't ideal, what is really needed is a "contains" however sadly a contains search is something that is not currently supported.

**Reply by PGLange** [reply]
hey @mike12345567
thanks for the fast reply. Sadly "Starts with" doesn't work either. 
Any other workaorund?

**Reply by mike12345567** [reply]
Hi @PGLange sadly not that I can think of, the format of your data looks quite difficult to match against, you would need something very custom to re-format the data before indexing it in the search engine, which isn't something we support right now.

In the near future we will be supporting contains searches within an SQL environment (connected to a Postgres/MySQL DB) in which case this would work, but for the time being I don't think theres anyway we can make CouchDB do this, it has certain limitations for performance reasons.

**Reply by PGLange** [reply]
The Field is just a workaround. In my apps there are two dynamical parties, each of them can take both roles. 
Is there a way to handle conditionals? If userA has role A then do stuff?

**Reply by mike12345567** [reply]
At the moment no, in the very near future we hope to put some functionality in place which will allow conditionally setting some properties of the UI using bindings (e.g. show/hide), the issue for this is here: https://github.com/Budibase/budibase/issues/1644

**Reply by PGLange** [reply]
Hey @mike12345567  Is there a way to realise an OR at the moment?

**Reply by mike12345567** [reply]
Sorry to be the bearer of bad news but no this isn't something that we have implemented yet, this is in the pipeline but we wanted to get an initial version of this out for people to play with and then flesh out some more advanced features afterwards.

**Reply by PGLange** [reply]
@mike12345567  My Idea now is to work with handlebars and an embed. lets see if that could work out. This is kind of a hacky thought - 
But I'm not used to Handlebars. Can you explain me how to cascade Handlebar blocks like {{#each}lala{{else}}{{/each}} and {{#and field_one field_two}}lala{{else}}{{/and}}?

**Reply by mike12345567** [reply]
I would recommend having a look [here](https://handlebarsjs.com/guide/block-helpers.html#conditionals) for help with handlebars conditionals, i'm not exactly sure what you mean by "cascading" but the idea with handlebars block helpers is that you can insert strings based on conditions, the most common use case for something like this is:
```
{{#if product}}
{{ product.name }}
{{else}}
No product found!
{{/if}}
```
You can chain these togethers like `{{#if ... }}{{#if ... }}{{/if}}{{/if}}` to check multiple different properties. To actually perform conditional logic, say the product price is greater than 100, you need to use helpers, e.g. `{{#if gte product.price 100}}More than 300{{/if}}`.

**Reply by PGLange** [reply]
Hey @mike12345567 your support is great - thanks a thousand times!
I'm trying to get the current User Object from with in a data Provider, but all paragraphs are empty, How can I access the CurrentUser object? f.e. {{Current User.email}}

**Reply by mike12345567** [reply]
Hi @PGLange no problem, heres a quick example of this:
![image](https://user-images.githubusercontent.com/4407001/123830611-5cc90180-d8fb-11eb-9360-4bd023e61b54.png)

**Reply by PGLange** [reply]
hi @mike12345567  Hm, I can feel I missed sth, ....
![like2](https://user-images.githubusercontent.com/86222990/123831221-65720580-d904-11eb-8384-ff6e3e88413e.PNG)

**Reply by mike12345567** [reply]
hmm I don't think the `""` are required around your `Current User.Company` - this may just be confusing the handlebars engine.

**Reply by PGLange** [reply]
![like1](https://user-images.githubusercontent.com/86222990/123835139-7ae92e80-d908-11eb-89a7-d80abdc786ec.PNG) I#m the third user in the list ...
![like2](https://user-images.githubusercontent.com/86222990/123835597-0793ec80-d909-11eb-93c7-5ac754abab62.PNG)


Hey @mike12345567  actually my Current User object just picks the first user - not the acutal logged in user - so no wonder that I couldn't get along with my logic - can you approve that? That'd be a solid bug :) ....

**Reply by mike12345567** [reply]
Ah @PGLange sorry this is specific to the preview, this is a little issue that we haven't quite found a solution to, when you're building in the builder preview some elements of the system aren't exactly as they are when the app is actually running. I'll loop in @aptkingston as he might have more to say on this topic.

When you're testing things like this you'd be better opening the proper preview of the app itself and reloading it to see the changes. 

Edit: you open this with the play symbol in the top right:
![image](https://user-images.githubusercontent.com/4407001/123835442-71f45f00-d900-11eb-9334-614675a02246.png)

**Reply by aptkingston** [reply]
> 
> 
> Ah @PGLange sorry this is specific to the preview, this is a little issue that we haven't quite found a solution to, when you're building in the builder preview some elements of the system aren't exactly as they are when the app is actually running. I'll loop in @aptkingston as he might have more to say on this topic.
> 
> When you're testing things like this you'd be better opening the proper preview of the app itself and reloading it to see the changes.
> 
> Edit: you open this with the play symbol in the top right:
> ![image](https://user-images.githubusercontent.com/4407001/123835442-71f45f00-d900-11eb-9334-614675a02246.png)

You're right that the builder preview shows the first user, rather than the current user. Although I just did some testing and I think we could actually change that now, due to the auth changes. I tried it there and it correctly identifes the current user from the preview so we could actually implement that now!

@PGLange if you use a repeater inside your data provider you'll be able to simplify the handlebars logic down a bit - it basically does the looping over each row for you.

**Reply by PGLange** [reply]
hey @ll,
id works! A solution for a logical or -as long as not implemented- is to use the Dataprovider to have (dynamical) filter options and then use an embed in the repeater (forms not tested jet). There {{#or dataProvider.param1 dataProvider.param2 Current User.email}}All the content here{{else}}Nothing needed here{{/or}}. Works out pretty well! THANK YOU GUYS!! <3

**Reply by PGLange** [reply]
Hi @aptkingston,
it worked fine ... and then ... don't know why since the update this simple {{#or}} doesn't work anymore. 
I can't figure out why. That's extremly frustrating - I was in the customer tests and now it fails totaly. 
My time is up - budibase won't be used, as the app doesn't work.

**Reply by aptkingston** [reply]
Hi @PGLange. I'm sorry to hear that. I'm surprised that updating broke the handlebars expression because I don't believe we made any changes to that in our latest release. I'm happy to help you get things working again if you'd like. It would be really helpful to get a better idea of what you're trying to accomplish with your app, because there may be a much easier way to achieve your goals than using complex handlebars statements to handle your logic.

**Reply by PGLange** [reply]
hi @aptkingston,
 {{#or dataProvider.param1 dataProvider.param2 Current User.email}}All the content here{{else}}Nothing needed here{{/or}}
Was the only logic I needed - at the end. 
I can't proove that the update is responsible for the broken app, maybe something else "broke".
The Update is the only thing I changed, execept deleting rows in the database. 
Sorry for all! 
And thank you for your offer but I got pulled for failure.
